File: Tang/Tang/Content/Module/Login/ViewCtrl/TangLoginViewCtrl.m
//
//  TangLoginViewCtrl.m
//  Tang
//
//  Created by wwwbbat on 2017/9/16.
//  Copyright © 2017年 tiny. All rights reserved.
//

#import "TangLoginViewCtrl.h"
#import "TFHpple.h"
#import "TangSession.h"
#import "BingoCache.h"
#import <SDWebImage/UIImageView+WebCache.h>

static NSString *const kTangLoginBackgroundImageCacheKey = @"login_background_img";

@interface TangLoginViewCtrl ()
@property (weak, nonatomic) IBOutlet UIImageView *bgImageView;
@property (weak, nonatomic) IBOutlet UIButton *oauthBtn;
@property (weak, nonatomic) IBOutlet UILabel *alertInfoLabel;
@property (weak, nonatomic) IBOutlet UILabel *postInfoLabel;

@end

@implementation TangLoginViewCtrl

- (void)viewDidLoad
{
    [super viewDidLoad];
    

    NSString *tmp = self.alertInfoLabel.text;
    self.alertInfoLabel.text = [NSString stringWithFormat:tmp,AppDisplayName];
    NSString *cachedImg = (NSString *)[APPCACHE objectForKey:kTangLoginBackgroundImageCacheKey];
    [self.bgImageView sd_setImageWithURL:[NSURL URLWithString:cachedImg]];
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(logined:) name:kTangSessionUserDidLoginedNotification object:nil];
    
    [self loadLoginPageInfo];
}

- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)loadLoginPageInfo
{
    RunOnGlobalQueue(^{
        NSString *url = @"https://www.tumblr.com";
        NSError *error;
        NSData *data = [NSData dataWithContentsOfURL:[NSURL URLWithString:url] options:NSDataReadingMappedIfSafe error:&error];
        if (!data || data.length < 10 || error ) {
            return;
        }
        TFHpple *root = [TFHpple hppleWithHTMLData:data];
        NSArray *results = [root searchWithXPathQuery:@"//img[@id='fullscreen_post_image']"];
        TFHppleElement *imgNode = results.firstObject;
        NSString *imgSrc = [imgNode objectForKey:@"src"];
        if (imgSrc) {
            NSLog(@"background image url:%@",imgSrc);
            RunOnMainQueue(^{
                [APPCACHE setObject:imgSrc forKey:kTangLoginBackgroundImageCacheKey];
                SDWebImageOptions options = SDWebImageProgressiveDownload | SDWebImageContinueInBackground | SDWebImageRetryFailed;
                [self.bgImageView sd_setImageWithURL:[NSURL URLWithString:imgSrc] placeholderImage:self.bgImageView.image options:options];
            });
        }
        
        //post info
        TFHppleElement *postInfoNode = [root searchWithXPathQuery:@"//div[@class='post_info just_tumblelog']"].firstObject;
        NSString *postinfo = [postInfoNode objectForKey:@"data-tumblelog"];
        if (postinfo) {
            RunOnMainQueue(^{
                self.postInfoLabel.text = [NSString stringWithFormat:@"图片由 %@ 发布",postinfo];
            });
        }
    });
}

- (void)logined:(NSNotification *)noti
{
    if (self.didLogined) {
        self.didLogined();
    }
}

- (IBAction)OAuthStart:(id)sender
{
    [API requestOAuth];
}

@end

File: Tang/Tang/AppCode/Module/Login/Ctrls/TangLoginViewCtrl.swift
//
//  TangLoginViewCtrl.swift
//  Tang
//
//  Created by Jay on 2018/2/8.
//  Copyright © 2018年 tiny. All rights reserved.
//

import UIKit
import SDWebImage

class TangLoginViewCtrl: YGBaseViewCtrl {
    
    private static let kTangLoginBackgroundImageCacheKey = "login_background_img"
    private static let kTangLoginPage = "https://www.tumblr.com"
    
    var didLogined : (()->Void)?
    
    @IBOutlet var bgImageView : UIImageView!
    @IBOutlet var oauthBtn : UIButton!
    @IBOutlet var alertInfoLabel : UILabel!
    @IBOutlet var postInfoLabel : UILabel!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        alertInfoLabel.text = String.init(format: alertInfoLabel.text!, AppDisplayName)
        let cachedImage = Cache.appCache().object(forKey: TangLoginViewCtrl.kTangLoginBackgroundImageCacheKey) as? String
        bgImageView.sd_setImage(with: URL.init(string: (cachedImage ?? "")), completed: nil)
        
        NotificationCenter.default.addObserver(self, selector: #selector(logined(_:)), name: kTangSessionUserDidLoginedNotification, object: nil)
        
        loadLoginPageInfo()
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self)
    }
    
    @objc private func logined(_ noti : NSNotification) {
        if let block = self.didLogined {
            block()
        }
    }
    
    private func loadLoginPageInfo() {
        RunOnGlobal {
            let url = URL.init(string: TangLoginViewCtrl.kTangLoginPage)
            guard let data = try? Data.init(contentsOf: url!, options: .mappedIfSafe), data.count > 10 else{
                return
            }
            
            let root = TFHpple.init(htmlData: data)
            
            // img info
            let imgNode = root?.search(withXPathQuery: "//img[@id='fullscreen_post_image']").first as? TFHppleElement
            if let imgSrc = imgNode?.object(forKey: "src") {
                print("login background image src : ",imgSrc)
                RunOnMain {
                    Cache.appCache().setObject(imgSrc as NSString, forKey: TangLoginViewCtrl.kTangLoginBackgroundImageCacheKey)
                    let options : SDWebImageOptions = [.progressiveDownload,.continueInBackground,.retryFailed]
                    self.bgImageView.sd_setImage(with: URL.init(string: imgSrc), placeholderImage: nil, options: options, completed: nil)
                }
            }
            
            // post info
            let postInfoNode = root?.search(withXPathQuery: "//div[@class='post_info just_tumblelog']").first as? TFHppleElement
            if let postInfo = postInfoNode?.object(forKey: "data-tumblelog") {
                RunOnMain {
                    self.postInfoLabel.text = String.init(format: "图片由 %@ 发布", postInfo)
                }
            }
        }
    }
    
    @IBAction func OAuthStart(_ sender : Any){
        API.instance.OAuth()
    }
}
